### PR TITLE
fixed issue that calloc cause COW pagefault

### PR DESCRIPTION
**Pull request by jinjunzh**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- HashTableAllocator call calloc to pre-fault the pages so that page faults don't impact performance during hash table resizing, but jemalloc use “MADV_DONTNEED” to clear memory. The problem is that “MADV_DONTNEED” do zero page which  would cause COW page fault if the address region is later touched. We see in certain cases when page fault rate is very high(>500K), the query performance would drop. After undefine JEMALLOC_PURGE_MADVISE_DONTNEED_ZEROS, it can boost QPS 1.2x.
test cases: [ontime](https://clickhouse.com/docs/en/getting-started/example-datasets/ontime) Q8 and Q10

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
This is an automated comment for commit 6ce6e549aca38981e8a7e9e4d25b4b7ac935b07e with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53267/6ce6e549aca38981e8a7e9e4d25b4b7ac935b07e/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by nickitat** [comment]
perf tests look cool, but check the `MemoryResident` metric (

**Comment by jinjunzh** [comment]
> MemoryResident

@nickitat Where is MemoryResident metric in check list?

**Comment by jinjunzh** [comment]
I changed to set Transparent hugepage (THP) mode:always according to below function in JEMALLOC.
./contrib/jemalloc/src/ehooks.c
`ehooks_default_zero_impl(void *addr, size_t size) {
	/*
	 * By default, we try to zero out memory using OS-provided demand-zeroed
	 * pages.  If the user has specifically requested hugepages, though, we
	 * don't want to purge in the middle of a hugepage (which would break it
	 * up), so we act conservatively and use memset.
	 */
	bool needs_memset = true;
	if (opt_thp != thp_mode_always) {
		needs_memset = pages_purge_forced(addr, size);
	}
	if (needs_memset) {
		memset(addr, 0, size);
	}
}`

**Comment by jinjunzh** [comment]
@nickitat I created below slides to summarize this issue status and 3 types of patches, please look into it and see if it helps:

image


image

**Comment by nickitat** [comment]
> > MemoryResident
> 
> @nickitat Where is MemoryResident metric in check list?

when you open the report at the very bottom you will see a table with different metrics including `MemoryResident` if it changed significantly

**Comment by nickitat** [comment]
1. we do pre-fault by passing `MAP_POPULATE` to [mmap](https://github.com/ClickHouse/ClickHouse/blob/master/src/Common/Allocator.h#L208-L219). it is done only for big allocations, that is why probably your initial patch improved queries with relatively small hash tables. this logic is about to be removed in #52792 and we will also have pre-fault for smaller allocations (see [this report](https://s3.amazonaws.com/clickhouse-test-reports/52792/b0e23c49847d8dd7b5b0734f5ddfd0463445c6bf/performance_comparison_[4_4]/report.html)), so this case seems to be also covered
2. afaik nor `calloc`, nor `memset(0)` are reliable ways to get populated pages, because they will instead just point all memory to a cow zero-page
3. unfortunately we cannot set thp support in jemalloc to "always enabled" either, because currently THP support in jemalloc is known to be poor and it will deteriorate performance on real workloads because of memory fragmentation. if you interested in the work around enabling THP in CH, pls see [this](https://github.com/ClickHouse/ClickHouse/issues/52608) issue

**Comment by jinjunzh** [comment]
@nickitat This PR: [Merge 51686](https://github.com/ClickHouse/ClickHouse/pull/52792) looks not be able to cover current case because its threshold size of pre-fault is defined as 16MB, while allocated size here for FixedHashTable is about 500KB~. So I submitted new commit to do pre-fault with MADV_POPULATE_WRITE after FixedHashTable allocation.

File: src/Common/Allocator.h [hunk 1]
```diff
@@ -144,6 +144,27 @@ class Allocator
...
+        if (len_ < page_size) /// Rounded address should be still within [buf, buf + len).
+            return;
+
+        auto [buf, len] = adjustToPageSize(buf_, len_, page_size);
+        if (auto res = ::madvise(buf, len, MADV_POPULATE_WRITE); res < 0)
+            throw DB::Exception(DB::ErrorCodes::BAD_ARGUMENTS,
+                                "Attempt to populate pages failed: {} (EINVAL is expected for kernels < 5.14)",
+                                errnoToString(res));
+    }
+
```
Comment (nickitat) [inline src/Common/Allocator.h:167]:
```suggestion
```

**Comment by nickitat** [comment]
@jinjunzh sorry for the mess, I've fixed compilation. let's look into the perf tests results

**Comment by azat** [comment]
I think that this pagefaults can be a problem for `FixedHashTable`, and in my understanding it should not increase the memory footprint, so it is strange that RSS increased (though I did not look by myself).

@jinjunzh question about your comment - https://github.com/ClickHouse/ClickHouse/pull/53267#issuecomment-1678343361, I don't see such code in jemalloc, in particular there is `extent_recycle` in `pages.c` (because it is in `extend.c`, and anyway the code there does not looks like the code from screenshot), can you post a link to the github to the code that you are referencing?

**Comment by azat** [comment]
>afaik nor calloc, nor memset(0) are reliable ways to get populated pages, because they will instead just point all memory to a cow zero-page

@nickitat interesting, I always thought that `memset(p, 0, size)` SHOULD populate pages, any reference?

**Comment by jinjunzh** [comment]
@azat 
Please refer to function definition here: [extent_recycle](https://github.com/ClickHouse/jemalloc/blob/e6891d9746143bf2cf617493d880ba5a0b9a3efd/src/extent.c)

**Comment by azat** [comment]
> About code section from screenshot, please refer to function definition here: [extent_recycle](https://github.com/ClickHouse/jemalloc/blob/e6891d9746143bf2cf617493d880ba5a0b9a3efd/src/extent.c)

Thanks, this repository is not used actually (it left only for previous releases), ClickHouse uses official jemalloc now - https://github.com/ClickHouse/ClickHouse/blob/ddf8ffb51c2fc55e4d4f0aee29d30689f5e583f7/.gitmodules#L33

But anyway right now the code is the same (more or less), I was just curious why lines did not matched.

**Comment by nickitat** [comment]
don't see any improved queries. thanks for all the work done anyway.

**Pull request closed without merging**